pkgs/cli_util/lib/src/base_directories.dart

```diff
@@ -0,0 +1,274 @@
+  /// nested in [cacheHome].
+  ///
+  /// Throws an [EnvironmentNotFoundException] if the necessary environment
+  /// variables are undefined.
+  String runtimeHome() =>
+      path.join(_baseDirectory(_XdgBaseDirectoryType.runtime), tool);
+
+  /// The path of the directory where [tool] can place its state.
+  ///
+  /// The state directory is likely not backed up or synchronized accross
+  /// devices by the OS. For data that may be backed up and synchronized, use
+  /// [dataHome] instead.
+  ///
+  /// This is a location appropriate for storing data which is either not
+  /// important enougn, not small enough, or not portable enough to store in
+  /// [dataHome]. For example: logs and indices.
+  ///
+  /// The directory location depends on the current [Platform.operatingSystem]:
+  /// - on **Windows**:
+  ///   - `%LOCALAPPDATA%\<tool>`
+  /// - on **Mac OS**:
+  ///   - `$HOME/Library/Application Support/<tool>`
+  /// - on **Linux**:
+  ///   - `$XDG_STATE_HOME/<tool>` if `$XDG_STATE_HOME` is defined, and
+  ///   - `$HOME/.local/state/<tool>` otherwise.
+  ///
+  /// The directory won't be created, the method merely returns the recommended
+  /// location.
+  ///
+  /// On some platforms, this path may overlap with [cacheHome], and
+  /// [runtimeHome].
+  ///
+  /// Throws an [EnvironmentNotFoundException] if the necessary environment
+  /// variables are undefined.
+  String stateHome() =>
+      path.join(_baseDirectory(_XdgBaseDirectoryType.state), tool);
+
+  String _baseDirectory(_XdgBaseDirectoryType dir) {
+    if (Platform.isWindows) {
+      return _baseDirectoryWindows(dir);
+    }
+    if (Platform.isMacOS) {
+      return _baseDirectoryMacOs(dir);
+    }
+    if (Platform.isLinux) {
+      return _baseDirectoryLinux(dir);
+    }
+    throw UnsupportedError('Unsupported platform: ${Platform.operatingSystem}');
+  }
+
+  String _baseDirectoryWindows(_XdgBaseDirectoryType dir) => switch (dir) {
+        _XdgBaseDirectoryType.config ||
+        _XdgBaseDirectoryType.data =>
+          _requireEnv('APPDATA'),
+        _XdgBaseDirectoryType.cache ||
+        _XdgBaseDirectoryType.runtime ||
+        _XdgBaseDirectoryType.state =>
+          _requireEnv('LOCALAPPDATA'),
+      };
+
+  String _baseDirectoryMacOs(_XdgBaseDirectoryType dir) => switch (dir) {
+        _XdgBaseDirectoryType.config =>
+          path.join(_home, 'Library', 'Preferences'),
+        _XdgBaseDirectoryType.data ||
+        _XdgBaseDirectoryType.state =>
+          path.join(_home, 'Library', 'Application Support'),
+        _XdgBaseDirectoryType.cache => path.join(_home, 'Library', 'Caches'),
+        _XdgBaseDirectoryType.runtime =>
+          // https://stackoverflow.com/a/76799489
+          path.join(_home, 'Library', 'Caches', 'TemporaryItems'),
+      };
+
+  String _baseDirectoryLinux(_XdgBaseDirectoryType dir) {
+    if (Platform.isLinux) {
+      final xdgEnv = switch (dir) {
+        _XdgBaseDirectoryType.config => 'XDG_CONFIG_HOME',
```

Comment:
Instance variable on enum: `final String _xdgKey;, insitialized to these values, then just
```dart
  final envVar = _environment[dir._xdgKey];
```
(Never switch over an enum to choose a fixed value. Put it on the enum.)

Reply:
If the fields on the enum are named `windowsThis` and `macOSThat`, I think that's separation enough. 

This enum is private, it exists only to enumerate the district cases that this code distinguishes. Putting _how_ the library distinguishes into the enum itself localizes the parameterization.
If we ever add a sixth kind of base directory, we would only need to add one enum entry, not an entry in four different switches.
(That's one of my measures for whether data is structured well: if you add one more concept, in how many places will you need to add an extra case.)